### PR TITLE
chore(backup): restrict state restore members to inventory

### LINKED ISSUE #2562: security(backup): restrict state restore members to inventory
# Implementation Plan

## Objective

Harden tier-3 restore so only uniquely named regular files beneath `INVENTORY` can be restored. Validate the complete archive and manifest before reading state payloads or writing files, preventing self-consistent archives from overwriting unauthorized repository files such as `pyproject.toml` while preserving legitimate backup, verify, and restore behavior.

## Approach

- Keep the implementation in the stdlib-only `scripts/backup_state.py`; ADR-0013 requires disaster-recovery tooling to run without `hephaestus` or optional dependencies (`docs/adr/0013-backup-and-disaster-recovery-policy.md:52-61`).
- Centralize archive-name validation in one `_canonical_member_name()` helper:
  - Reject empty, absolute, backslash-containing, NUL-containing, and `..` paths.
  - Normalize safe POSIX paths before duplicate detection and authorization.
  - Use a separate `_inventory_prefix_for()` helper only for inventory ownership, avoiding duplicated path-safety checks.
- Enumerate `TarFile.getmembers()` once before reading payloads:
  - Permit exactly one regular canonical `manifest.json`.
  - Require every other member to be a regular file strictly below an `INVENTORY` prefix.
  - Reject exact and normalization-equivalent duplicates, links, directories, devices, and unknown paths.
- Read archive content through the validated `TarInfo` objects rather than name-based `extractfile()` calls. This avoids duplicate-name lookup ambiguity.
- Parse the manifest fail-closed:
  - Convert invalid UTF-8, invalid JSON, duplicate JSON object keys, wrong top-level shape, invalid `members`, and malformed `sha256`/`size` metadata into `RestoreError`.
  - Normalize and inventory-authorize manifest keys with the same canonical helper.
  - Require an exact one-to-one set match between manifest entries and validated data members.
- Migrate both consumers of `_read_manifest()`:
  - `cmd_restore()` receives the manifest through `_validated_restore_members()`.
  - `cmd_verify()` also uses `_validated_restore_members()`, preserving return code `0` for valid archives and returning `1` with a `FAIL` message for structurally invalid archives.
- Resolve each restore destination beneath both the repository and its lexical inventory root before reading its payload. This prevents pre-existing symlinks inside an authorized inventory path from redirecting writes elsewhere.
- Retain the verify-then-write transaction: validate structure and destinations, read and verify all size/digest metadata into memory, then perform writes only after every member succeeds.
- Existing precedents support the design:
  - `hephaestus/prompts/catalog.py:95-105` validates untrusted names with `PurePosixPath`.
  - `hephaestus/automation/pipeline/worker_pool.py:725-761` enumerates tar members, rejects unsafe types, and reads only validated `TarInfo` objects.
  - `scripts/backup_state.py:138-168` and `scripts/backup_state.py:171-214` are the only verify/restore consumers; `tests/unit/scripts/test_backup_state.py` is their established behavior-test boundary.
- No dependency, archive migration, or rollout is required. Archives produced by the current `cmd_backup()` already contain one regular manifest and regular files under `build/.issue_implementer/`.

## Files to Create

_None._

### `scripts/backup_state.py`

Import `PurePosixPath` at `scripts/backup_state.py:41` and add centralized canonicalization, inventory authorization, and strict manifest parsing around the current `_read_manifest()` at `scripts/backup_state.py:111`:

```python
from pathlib import Path, PurePosixPath


def _canonical_member_name(name: str) -> str:
    """Return one safe canonical POSIX archive-member name."""
    if not name or "\x00" in name or "\\" in name:
        raise RestoreError(f"unsafe archive member path: {name!r}")

    path = PurePosixPath(name)
    if path.is_absolute() or ".." in path.parts:
        raise RestoreError(f"unsafe archive member path: {name!r}")

    normalized = path.as_posix()
    if normalized == ".":
        raise RestoreError(f"unsafe archive member path: {name!r}")
    return normalized


def _inventory_prefix_for(normalized: str) -> str:
    """Return the inventory prefix that strictly owns a canonical member."""
    path = PurePosixPath(normalized)
    for prefix in INVENTORY:
        inventory = PurePosixPath(prefix)
        if path != inventory and path.is_relative_to(inventory):
            return prefix
    raise RestoreError(f"archive member is outside inventory: {normalized!r}")


def _unique_json_object(pairs: list[tuple[str, object]]) -> dict[str, object]:
    """Build a JSON object while rejecting duplicate keys."""
    result: dict[str, object] = {}
    for key, value in pairs:
        if key in result:
            raise RestoreError(f"duplicate manifest key: {key!r}")
        result[key] = value
    return result


def _read_manifest(
    tar: tarfile.TarFile,
    member: tarfile.TarInfo,
) -> dict[str, dict[str, object]]:
    """Read and validate the manifest from its validated archive member."""
    extracted = tar.extractfile(member)
    if extracted is None:
        raise RestoreError(f"archive has no readable {MANIFEST_NAME}")

    try:
        document = json.loads(
            extracted.read().decode("utf-8"),
            object_pairs_hook=_unique_json_object,
        )
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise RestoreError(f"invalid {MANIFEST_NAME}: {exc}") from exc

    if not isinstance(document, dict) or set(document) != {"members"}:
        raise RestoreError(f"{MANIFEST_NAME} must contain exactly a members object")
    raw_members = document["members"]
    if not isinstance(raw_members, dict):
        raise RestoreError(f"{MANIFEST_NAME} members must be an object")

    members: dict[str, dict[str, object]] = {}
    for name, metadata in raw_members.items():
        if not isinstance(metadata, dict) or set(metadata) != {"sha256", "size"}:
            raise RestoreError(f"invalid manifest metadata for {name!r}")

        digest = metadata["sha256"]
        size = metadata["size"]
        if not isinstance(digest, str) or len(digest) != 64:
            raise RestoreError(f"invalid SHA-256 for {name!r}")
        try:
            int(digest, 16)
        except ValueError as exc:
            raise RestoreError(f"invalid SHA-256 for {name!r}") from exc
        if type(size) is not int or size < 0:
            raise RestoreError(f"invalid size for {name!r}")

        members[name] = metadata
    return members
```

Add a complete archive-index validator before `cmd_verify()` at `scripts/backup_state.py:138`. All raw path checks flow through `_canonical_member_name()`:

```python
def _validated_restore_members(
    tar: tarfile.TarFile,
) -> tuple[
    dict[str, dict[str, object]],
    dict[str, tuple[tarfile.TarInfo, str]],
]:
    """Validate and bind the manifest and authorized regular data members."""
    manifest_member: tarfile.TarInfo | None = None
    archive_members: dict[str, tuple[tarfile.TarInfo, str]] = {}
    seen: set[str] = set()

    for member in tar.getmembers():
        normalized = _canonical_member_name(member.name)
        if normalized in seen:
            raise RestoreError(f"duplicate archive member: {normalized!r}")
        seen.add(normalized)

        if not member.isreg():
            raise RestoreError(f"archive member is not a regular file: {member.name!r}")

        if normalized == MANIFEST_NAME:
            manifest_member = member
            continue

        prefix = _inventory_prefix_for(normalized)
        archive_members[normalized] = (member, prefix)

    if manifest_member is None:
        raise RestoreError(f"archive is missing {MANIFEST_NAME}")

    raw_manifest = _read_manifest(tar, manifest_member)
    manifest: dict[str, dict[str, object]] = {}
    for raw_name, metadata in raw_manifest.items():
        normalized = _canonical_member_name(raw_name)
        _inventory_prefix_for(normalized)
        if normalized in manifest:
            raise RestoreError(f"duplicate normalized manifest member: {normalized!r}")
        manifest[normalized] = metadata

    if set(manifest) != set(archive_members):
        raise RestoreError("archive members do not exactly match the manifest")
    return manifest, archive_members
```

Migrate `cmd_verify()` at `scripts/backup_state.py:138-168` to the validated archive index, explicitly preserving its caller-visible integer result:

```python
def cmd_verify(archive: Path) -> int:
    """Verify archive structure, authorization, sizes, and digests read-only."""
    try:
        with tarfile.open(archive, "r:gz") as tar:
            manifest, archive_members = _validated_restore_members(tar)
            ok = True
            for name, metadata in sorted(manifest.items()):
                member, _ = archive_members[name]
                extracted = tar.extractfile(member)
                data = extracted.read() if extracted is not None else None
                if (
                    data is None
                    or len(data) != metadata["size"]
                    or _sha256_bytes(data) != metadata["sha256"]
                ):
                    print(f"FAIL {name} (content mismatch)")
                    ok = False
                else:
                    print(f"PASS {name}")
            return 0 if ok else 1
    except RestoreError as exc:
        print(f"FAIL archive ({exc})")
        return 1
```

Refactor `cmd_restore()` at `scripts/backup_state.py:171-214` to consume validated `TarInfo` objects and enforce both repository and inventory destination boundaries:

```python
with tarfile.open(archive, "r:gz") as tar:
    manifest, archive_members = _validated_restore_members(tar)

    if not force:
        for prefix in INVENTORY:
            base = repo_root / prefix
            if base.exists() and any(base.rglob("*")):
                raise RestoreError(
                    f"target {base} is not empty; pass force=True to overwrite"
                )

    repo_resolved = repo_root.resolve()
    staged: dict[Path, bytes] = {}
    for name, metadata in manifest.items():
        member, prefix = archive_members[name]
        inventory_root = repo_resolved / prefix
        dest = (repo_resolved / name).resolve()
        if not _is_within(repo_resolved, dest) or not _is_within(inventory_root, dest):
            raise RestoreError(f"refusing member outside inventory: {name!r}")

        extracted = tar.extractfile(member)
        data = extracted.read() if extracted is not None else None
        if data is None:
            raise RestoreError(f"member is unreadable: {name!r}")
        if len(data) != metadata["size"]:
            raise RestoreError(f"size mismatch for {name!r}; refusing restore")
        if _sha256_bytes(data) != metadata["sha256"]:
            raise RestoreError(f"digest mismatch for {name!r}; refusing restore")
        staged[dest] = data

    for dest, data in staged.items():
        dest.parent.mkdir(parents=True, exist_ok=True)
        dest.write_bytes(data)
```

Update the module and function docstrings at `scripts/backup_state.py:2-29`, `scripts/backup_state.py:138-147`, and `scripts/backup_state.py:171-184` to state that verification and restore reject invalid structure, unauthorized members, malformed manifests, and size/digest mismatches.

### `tests/unit/scripts/test_backup_state.py`

Add reusable explicit-archive construction helpers near `tests/unit/scripts/test_backup_state.py:110` so tests can create duplicate, non-regular, unauthorized, and malformed entries without touching live state:

```python
def _write_test_archive(
    path: Path,
    entries: list[tuple[tarfile.TarInfo, bytes | None]],
    manifest: bytes,
) -> Path:
    """Write an archive with caller-controlled members and manifest bytes."""
    import io

    with tarfile.open(path, "w:gz") as tar:
        manifest_info = tarfile.TarInfo(backup_state.MANIFEST_NAME)
        manifest_info.size = len(manifest)
        tar.addfile(manifest_info, io.BytesIO(manifest))
        for member, data in entries:
            stream = io.BytesIO(data) if data is not None else None
            tar.addfile(member, stream)
    return path
```

Add malicious overwrite coverage after the existing traversal regression at `tests/unit/scripts/test_backup_state.py:189-212`:

```python
def test_restore_rejects_in_repo_member_outside_inventory_without_writing(
    tmp_path: Path,
) -> None:
    """A digest-valid archive cannot overwrite an unauthorized repo file."""
    repo_root = tmp_path / "repo"
    repo_root.mkdir()
    protected = repo_root / "pyproject.toml"
    protected.write_bytes(b"SAFE")

    archive = _archive_with_regular_members(
        tmp_path / "evil.tar.gz",
        {"pyproject.toml": b"PWNED"},
    )

    with pytest.raises(backup_state.RestoreError):
        backup_state.cmd_restore(repo_root, archive, force=True)

    assert protected.read_bytes() == b"SAFE"
    assert not (repo_root / "build" / ".issue_implementer").exists()
```

Add behavior-level restore tests for absolute names, parent traversal, undeclared inventory members, missing members, exact duplicates, normalization-equivalent duplicates, directories, symbolic links, hard links, and devices. Every case must assert `RestoreError` and no created inventory file:

```python
@pytest.mark.parametrize(
    "case",
    [
        "absolute",
        "parent_traversal",
        "undeclared",
        "missing",
        "duplicate",
        "normalized_duplicate",
        "directory",
        "symbolic_link",
        "hard_link",
        "device",
    ],
)
def test_restore_rejects_invalid_archive_members(
    tmp_path: Path,
    case: str,
) -> None:
    """Unsafe, unknown, duplicate, and non-regular members fail closed."""
    repo_root = tmp_path / "repo"
    repo_root.mkdir()
    archive = _write_invalid_archive(tmp_path / f"{case}.tar.gz", case)

    with pytest.raises(backup_state.RestoreError):
        backup_state.cmd_restore(repo_root, archive, force=True)

    assert not (repo_root / "build" / ".issue_implementer").exists()
```

Add malformed-manifest coverage for invalid UTF-8/JSON, non-object roots, missing or non-object `members`, duplicate JSON keys, non-object metadata, invalid digest values, boolean/negative sizes, and normalized duplicate manifest paths:

```python
@pytest.mark.parametrize("manifest", MALFORMED_MANIFESTS)
def test_restore_rejects_malformed_manifest_as_restore_error(
    tmp_path: Path,
    manifest: bytes,
) -> None:
    """Malformed manifests use the fail-closed RestoreError boundary."""
    repo_root = tmp_path / "repo"
    repo_root.mkdir()
    archive = _write_test_archive(tmp_path / "bad.tar.gz", [], manifest)

    with pytest.raises(backup_state.RestoreError):
        backup_state.cmd_restore(repo_root, archive, force=True)

    assert not (repo_root / "build").exists()
```

Add a verify regression proving malformed archives return `1`, while retaining `test_verify_passes_on_untampered_archive()` at `tests/unit/scripts/test_backup_state.py:102-107` to guard the migrated helper contract:

```python
def test_verify_rejects_malformed_manifest(tmp_path: Path) -> None:
    """Verify reports structural failure instead of leaking a parsing exception."""
    archive = _write_test_archive(tmp_path / "bad.tar.gz", [], b"{not-json")
    assert backup_state.cmd_verify(archive) == 1
```

Retain the existing round-trip, tamper, forced overwrite, empty inventory, CLI verify, and traversal tests.

### `docs/runbooks/backup-restore.md`

Expand the fail-closed contract at `docs/runbooks/backup-restore.md:47-55` and update the verification description at `docs/runbooks/backup-restore.md:76-88`:

```markdown
- An archive may contain exactly one regular `manifest.json` plus regular files
  whose normalized paths are strictly beneath `build/.issue_implementer/`.
- Unknown, undeclared, missing, duplicate, absolute, path-traversal, directory,
  link, and device members are rejected before any state payload is read or
  written.
- Malformed manifests and invalid size or SHA-256 metadata are rejected.
- Every authorized member's declared size and SHA-256 digest are verified before
  anything is written; any failure leaves the repository untouched.
```

Document that `verify` applies the same archive-structure and inventory checks and returns `1` for structural, metadata, size, or digest failures.

## Implementation Order

1. Add controlled archive-building helpers and failing tests for unauthorized in-repository overwrite, invalid tar members, malformed manifests, and `cmd_verify()` compatibility.
2. Add `_canonical_member_name()` and `_inventory_prefix_for()` so path normalization and raw safety validation have one authority.
3. Harden `_read_manifest()` with validated `TarInfo` input, duplicate-key detection, shape checks, and metadata validation that consistently raises `RestoreError`.
4. Add `_validated_restore_members()` to validate the entire tar index, authorize inventory paths, and require exact archive/manifest membership.
5. Migrate `cmd_verify()` to `_validated_restore_members()` and retain its `0`/`1` return contract.
6. Migrate `cmd_restore()` to validated `TarInfo` reads, repository/inventory destination checks, size/digest verification, and write-after-complete-validation staging.
7. Update operator-facing docstrings and `docs/runbooks/backup-restore.md`.
8. Run the criterion-specific tests, complete backup/runbook suites, and Ruff checks.

## Verification

- Acceptance criterion: normalize and validate every archive member against allowed `INVENTORY` prefixes before staging or extraction.

```bash
uv run pytest tests/unit/scripts/test_backup_state.py::test_restore_rejects_invalid_archive_members tests/unit/scripts/test_backup_state.py::test_restore_rejects_in_repo_member_outside_inventory_without_writing -v
```

- Acceptance criterion: reject unknown, duplicate, non-regular, absolute, and traversal members.

```bash
uv run pytest tests/unit/scripts/test_backup_state.py::test_restore_rejects_invalid_archive_members tests/unit/scripts/test_backup_state.py::test_restore_rejects_path_traversal -v
```

- Acceptance criterion: cover malicious in-repository overwrite in addition to existing `../` traversal coverage.

```bash
uv run pytest tests/unit/scripts/test_backup_state.py::test_restore_rejects_in_repo_member_outside_inventory_without_writing tests/unit/scripts/test_backup_state.py::test_restore_rejects_path_traversal -v
```

- Acceptance criterion: preserve verified restore for legitimate tier-3 state.

```bash
uv run pytest tests/unit/scripts/test_backup_state.py::test_backup_restore_round_trip tests/unit/scripts/test_backup_state.py::test_restore_force_overwrites tests/unit/scripts/test_backup_state.py::test_verify_passes_on_untampered_archive -v
```

- Reviewer finding: preserve `cmd_verify()` after the `_read_manifest()` contract migration.

```bash
uv run pytest tests/unit/scripts/test_backup_state.py::test_verify_passes_on_untampered_archive tests/unit/scripts/test_backup_state.py::test_verify_detects_tampering tests/unit/scripts/test_backup_state.py::test_verify_rejects_malformed_manifest tests/unit/scripts/test_backup_state.py::test_cli_backup_then_verify -v
```

- Reviewer finding: malformed manifest failures remain inside the `RestoreError`/CLI boundary.

```bash
uv run pytest tests/unit/scripts/test_backup_state.py::test_restore_rejects_malformed_manifest_as_restore_error tests/unit/scripts/test_backup_state.py::test_cli_restore_refused_returns_two -v
```

- Full regression and documentation validation:

```bash
uv run pytest tests/unit/scripts/test_backup_state.py tests/unit/docs/test_backup_restore_runbook.py -v
```

- Static checks:

```bash
uv run ruff check scripts/backup_state.py tests/unit/scripts/test_backup_state.py
```

```bash
uv run ruff format --check scripts/backup_state.py tests/unit/scripts/test_backup_state.py
```

## Skills Used

- `athena:plan-issue` — applied the repository’s architecture-first, behavior-verifiable planning contract.
- `athena:advise` — the supplied team knowledge result contained no applicable durable skills or guidance.
- Prior team knowledge-base skills — none.

## Changes from Review

- Addressed the major `cmd_verify()` contract finding by explicitly migrating it to `_validated_restore_members()`, preserving its integer return behavior, and adding untampered, tampered, malformed, and CLI verification commands.
- Addressed the DRY finding by making `_canonical_member_name()` the sole authority for absolute, traversal, separator, empty-name, and normalization checks; `_inventory_prefix_for()` handles only ownership.
- Addressed the malformed-manifest finding by defining duplicate-key, decoding, JSON shape, members-map, digest, and size validation, converting failures to `RestoreError`, and adding parameterized restore plus verify coverage.
- Preserved the previously accepted requirement mapping for unauthorized `pyproject.toml` overwrite protection, invalid tar member types, exact manifest/archive membership, and legitimate tier-3 round-trip restore.

### PR DESCRIPTION
**Pull request by mvillmow**
## Summary
Implements the requested changes for issue #2562.

## Changes
See the PR diff for the full change set.

## Testing
Not run by the automation pipeline.

## Closes
Closes #2562

Generated by Hephaestus automation

### CODE REVIEW

File: scripts/backup_state.py [hunk 1]
```diff
@@ -101,29 +106,153 @@ def cmd_backup(repo_root: Path, output_dir: Path, timestamp: str) -> Path:
...
+    archive_members: dict[str, tuple[tarfile.TarInfo, str]] = {}
+    seen: set[str] = set()
+
+    for member in tar.getmembers():
+        normalized = _canonical_member_name(member.name)
+        if normalized in seen:
+            raise RestoreError(f"duplicate archive member: {normalized!r}")
+        seen.add(normalized)
+
+        if not member.isreg():
```
Comment (mvillmow) [inline scripts/backup_state.py:290]: [Review] `cmd_backup()` still enumerates `Path.is_file()` entries (which follow symlinks), while `tarfile` preserves symbolic and hard links by default. An inventory symlink or hard-linked state files can therefore produce a `SYMTYPE`/`LNKTYPE` member that this new check rejects. Backup reports success, but its output then fails both `verify` and `restore`—a POLA/recovery regression. Make backup guarantee regular output (reject links before success, or materialize approved regular bytes without following symlinks outside inventory) and add backup→verify coverage for symlink and hard-link inputs.

Comment (mvillmow) [reply scripts/backup_state.py:290]: [Response] Backup now rejects symlinks before archive creation, materializes hard links as regular tar members, and includes backup-to-verify regression coverage.
